src/System.Linq/tests/ConcatTests.cs

```diff
@@ -206,6 +206,16 @@ public void CountOfConcatIteratorShouldThrowExceptionOnIntegerOverflow()
             Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).Count());
             Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).Count());
             Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).Count());
+
+            // This applies to ToArray() and ToList() as well, which try to preallocate the exact size
+            // needed if all inputs are ICollections.
+            Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).ToArray());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).ToArray());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).ToArray());
+
+            Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).ToList());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).ToList());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).ToList());
```

Comment:
@stephentoub Yes, that's why I put them in this PR. We now take the Count into acct when calling `ToArray`, and pre-allocate a list of the correct size for `ToList` (Count in both cases throws for overflow). I'm sure a similar behavior change was introduced when `Concat` was changed from [this](https://github.com/dotnet/corefx/blob/release/1.0.0-rc1/src/System.Linq/src/System/Linq/Enumerable.cs#L1094-L1098) implementation to the one we have now, since then `Concat(...).ToList()` would have worked.